winusb.c

```diff
@@ -0,0 +1,156 @@
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <libopencm3/usb/usbd.h>
+#include "winusb.h"
+
+static struct winusb_compatible_id_descriptor winusb_wcid = {
+	.header = {
+		.dwLength = sizeof(struct winusb_compatible_id_descriptor_header) +
+					1 * sizeof(struct winusb_compatible_id_function_section),
+		.bcdVersion = WINUSB_BCD_VERSION,
+		.wIndex = WINUSB_REQ_GET_COMPATIBLE_ID_FEATURE_DESCRIPTOR,
+		.bNumSections = 1,
+		.reserved = { 0, 0, 0, 0, 0, 0, 0 },
+	},
+	.functions = {
+		{
+			.bInterfaceNumber = 2,
+			.reserved0 = { 1 },
+			.compatibleId = "WINUSB",
+			.subCompatibleId = "",
+			.reserved1 = { 0, 0, 0, 0, 0, 0}
+		},
+	}
+};
+
+static struct usb_string_descriptor winusb_string_descriptor = {
+  .bLength = sizeof(WINUSB_EXTRA_STRING) + sizeof(struct usb_string_descriptor),
+  .bDescriptorType = USB_DT_STRING,
+  .wData = WINUSB_EXTRA_STRING
+};
+
+static const struct winusb_extended_properties_descriptor guid = {
+	.header = {
+		.dwLength = sizeof(struct winusb_extended_properties_descriptor_header)
+					+ 1 * sizeof (struct winusb_extended_properties_feature_descriptor),
+		.bcdVersion = WINUSB_BCD_VERSION,
+		.wIndex = WINUSB_REQ_GET_EXTENDED_PROPERTIES_OS_FEATURE_DESCRIPTOR,
+		.wNumFeatures = 1,
+	},
+	.features = {
+		{
+			.dwLength = sizeof(struct winusb_extended_properties_feature_descriptor),
+			.dwPropertyDataType = WINUSB_EXTENDED_PROPERTIES_MULTISZ_DATA_TYPE,
+			.wNameLength = sizeof(u"DeviceInterfaceGUIDs"),
+			.name = u"DeviceInterfaceGUIDs",
+			.dwPropertyDataLength = (WINUSB_EXTENDED_PROPERTIES_GUIDS_SIZE_C),
+			.propertyData = WINUSB_EXTENDED_PROPERTIES_RANDOM_GUID,
+		},
+	}
+};
+
+static int winusb_descriptor_request(usbd_device *usbd_dev,
+					struct usb_setup_data *req,
+					uint8_t **buf, uint16_t *len,
+					usbd_control_complete_callback* complete) {
+	(void)complete;
+	(void)usbd_dev;
+
+	if ((req->bmRequestType & USB_REQ_TYPE_TYPE) != USB_REQ_TYPE_STANDARD) {
+		return USBD_REQ_NEXT_CALLBACK;
+	}
+
+	if (req->bRequest == USB_REQ_GET_DESCRIPTOR && req->wValue >> 8 == USB_DT_STRING) {
+		if ((req->wValue & 0xff) == WINUSB_EXTRA_STRING_INDEX) {
+			*buf = (uint8_t*)(&winusb_string_descriptor);
+			if (*len > winusb_string_descriptor.bLength) {
+				*len = winusb_string_descriptor.bLength;
+			}
+			return USBD_REQ_HANDLED;
+		}
+	}
+	return USBD_REQ_NEXT_CALLBACK;
+}
+
+static int winusb_control_vendor_request(usbd_device *usbd_dev,
+					struct usb_setup_data *req,
+					uint8_t **buf, uint16_t *len,
+					usbd_control_complete_callback* complete) {
+	(void)complete;
+	(void)usbd_dev;
+
+	if (req->bRequest != WINUSB_MS_VENDOR_CODE) {
+		return USBD_REQ_NEXT_CALLBACK;
+	}
+
+	int status = USBD_REQ_NOTSUPP;
+	if (((req->bmRequestType & USB_REQ_TYPE_RECIPIENT) == USB_REQ_TYPE_DEVICE) &&
+		(req->wIndex == WINUSB_REQ_GET_COMPATIBLE_ID_FEATURE_DESCRIPTOR)) {
+		*buf = (uint8_t*)(&winusb_wcid);
+		if (*len > winusb_wcid.header.dwLength) {
```

Comment:
Why do you need this check?

Reply:
Not sure (it's from the devanlai library), but this is basically what is done in libopencp3 on here and on similar lines

https://github.com/libopencm3/libopencm3/blob/master/lib/usb/usb_standard.c#L160

(this is basically the same, just written differently)

Reply:
Perhaps we ought to change https://github.com/trezor/trezor-mcu/blob/bc7c66aa763858fa751e8f6353a8e18a776da65c/firmware/usb.c#L322

Reply:
That should do the same thing as this; and again, this is also using similar style to libopencm3 code that uses the headers, not the structs. It's nicer though, yeah

Reply:
@karel-3d I'm suggesting that we change `usb.c` to match this (compare `*len` and use `dwLength`)

Reply:
Oh OK, I misread the comment. :)

I have cleaned up the code a bit so it now uses MIN so it is perhaps cleaner.

Reply:
I have added this logic to usb.c (both firmware and bootloader, although it probably won't matter there) and also usb21_standard.c, it should be in this PR also